apps/workshop/projects/BOBProtocol/helpers/vaults.ts

```diff
@@ -0,0 +1,192 @@
+import { Address, Pool, Token, PoolSupplied, SupportedChain } from '@defiyield/sandbox';
+import { VaultConfig, VAULTS, VAULT_TOKENS } from './config';
+import { Pair, OrderBook, getPairs, getOrderBookForTicker, getTvlFromOrderbook } from './api';
+
+const MGS_UNSOPPORTED_CHAIN = 'Unsupported chain';
+
+type TokenInPair = {
+  [symbol: string]: string;
+};
+
+type Orderbooks = {
+  [ticker: string]: OrderBook;
+};
+
+type VaultPerChain = {
+  [chainId: string]: Vault;
+};
+
+export class Vaults {
+  private static instance: Vaults;
+  private chain_vaults: VaultPerChain = {};
+
+  private constructor() {
+    this.chain_vaults = {};
+  }
+
+  public static getInstance(): Vaults {
+    if (!Vaults.instance) {
+      Vaults.instance = new Vaults();
+    }
+
+    return Vaults.instance;
+  }
+
+  public async getVault(chain: SupportedChain, axios: any, logger: any): Promise<Vault> {
+    if (!this.chain_vaults[chain]) {
+      this.chain_vaults[chain] = new Vault(chain, axios, logger);
+      await this.chain_vaults[chain].initVault();
+    }
+    return this.chain_vaults[chain];
+  }
+}
+
+function getVaultConfig(chain: SupportedChain): VaultConfig {
+  const currentVault = VAULTS[chain];
+  if (!currentVault) {
+    throw new Error(MGS_UNSOPPORTED_CHAIN);
+  }
+  return currentVault;
+}
+
+export class Vault {
+  private chain: SupportedChain = 'ethereum';
+  private id = '';
+  private base_url = '';
+  private tokens: TokenInPair = {};
+  private cached_orderbooks: Orderbooks = {};
+  private axios: any;
+  private logger: any;
+
+  public constructor(chain: SupportedChain, axios: any, logger: any) {
+    this.chain = chain;
+    this.axios = axios;
+    this.logger = logger;
+  }
+
+  private log_info(message: string) {
+    this.logger.info(message);
+  }
+
+  private log_warn(message: string) {
+    this.logger.warn(message);
+  }
+
+  private getTokenAddressBySymbol(symbol: string): Address | null {
+    // TODO: wait for BOBStats API update to get token addresses from there
```

Comment:
Is this will be updated in the scope of current PR?

Reply:
In order to get information about BobSwap pools for de.fi we re-use an existing API utilised initially to compose data for CoinGecko. This API does not contain token addresses since it is not required by CG. Our plan is to extend API in near future to refer to the tokens addresses so we will prepare update for de.fi simplifying the logic to provide pools' tokens.